testdata/generate.go

```diff
@@ -0,0 +1,245 @@
+	}); err != nil {
+		return "", "", err
+	}
+
+	return caCertPEM.String(), caPrivateKeyPEM.String(), nil
+}
+
+// removePemHeaders strips the PEM headers from a certificate so it can be used in an Ownership Voucher.
+func removePemHeaders(pemBlock string) string {
+	pemBlock = strings.TrimPrefix(pemBlock, "-----BEGIN CERTIFICATE-----\n")
+	pemBlock = strings.TrimSuffix(pemBlock, "\n-----END CERTIFICATE-----\n")
+	return pemBlock
+}
+
+// newOwnershipVoucher creates an OV for the device serial which is signed by the vendor's CA.
+func newOwnershipVoucher(serial string, pdcPem []byte, vendorCACert *x509.Certificate, vendorCAPriv *rsa.PrivateKey) (string, error) {
+	currentTime := time.Now()
+	ov := OwnershipVoucher{
+		OV: OwnershipVoucherInner{
+			CreatedOn:        currentTime.String(),
+			ExpiresOn:        currentTime.Add(caExpiry).String(),
+			SerialNumber:     serial,
+			PinnedDomainCert: removePemHeaders(string(pdcPem)),
+		},
+	}
+
+	ovBytes, err := json.Marshal(ov)
+	if err != nil {
+		return "", err
+	}
+
+	signedMessage, err := pkcs7.NewSignedData(ovBytes)
+	if err != nil {
+		return "", err
+	}
+	signedMessage.SetDigestAlgorithm(pkcs7.OIDDigestAlgorithmSHA256)
+	signedMessage.SetEncryptionAlgorithm(pkcs7.OIDEncryptionAlgorithmRSA)
+
+	err = signedMessage.AddSigner(vendorCACert, vendorCAPriv, pkcs7.SignerInfoConfig{})
+	if err != nil {
+		return "", err
+	}
+
+	signedBytes, err := signedMessage.Finish()
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(signedBytes), nil
+}
+
+// writeFile writes the contents to a new file in the current directory.
+func writeFile(contents []byte, filename string) error {
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	b, err := f.Write(contents)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Wrote %d bytes to %v\n", b, filename)
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	if *vendor == "" {
+		log.Exitf("vendor flag must be set")
+	}
+	if *owner == "" {
+		log.Exitf("owner flag must be set")
+	}
+	serials := strings.Split(*controlCardSerials, ",")
+	if len(serials) == 0 {
+		log.Exitf("no control card serial numbers provided")
+	}
+
+	// Generate vendor CA
+	fmt.Printf("Generating %v Root CA cert and private key\n", *vendor)
+	vendorCAPub, vendorCAPriv, err := newCertificateAuthority("Manufacturer Root CA", *vendor)
+	if err != nil {
+		log.Exitf("unable to generate vendor CA: %v", err)
+	}
+	if err := writeFile([]byte(vendorCAPub), "vendorca_pub.pem"); err != nil {
+		log.Exit(err)
+	}
+	if err := writeFile([]byte(vendorCAPriv), "vendorca_priv.pem"); err != nil {
+		log.Exit(err)
+	}
+
+	//Generate PDC.
+	fmt.Printf("Generating %v PDC cert and private key\n", *owner)
+	pdc, pdcPriv, err := newCertificateAuthority("Device Owner PDC", *owner)
+	if err != nil {
+		log.Exitf("unable to generate PDC: %v", err)
+	}
+	if err := writeFile([]byte(pdc), "pdc_pub.pem"); err != nil {
```

Comment:
so given you created these functions having to then "cohearse" things into other formats seems counter to design